### PR TITLE
Infer output predicates for inner joins in equality inference

### PR DESCRIPTION
**Pull request by sopel39**
There are more possible improvements in join equality inference:
* for outer joins outer side we could infer via inner side symbols (with `OR inner side symbols null` condition)
* for outer joins inner side we could infer via outer side symbols with (with `OR inner side symbols null` condition)
* for outer joins filter conditions we could produce predicate (with `OR inner side symbols null` condition)

Additionally equality inference `rewrite` mechianism could be improved so that expressions like:
```
p1 OR (p2 AND p3 AND p4)
```
are rewritten (expanded) into:
```
p1 OR (p2 AND p4)
```
when `p3` cannot be rewritten. Currently `rewrite` would just return `null`

### CODE REVIEW

File: presto-main/src/test/java/com/facebook/presto/sql/planner/TestEffectivePredicateExtractor.java [hunk 1]
```diff
@@ -458,7 +458,102 @@ public void testInnerJoin()
                         equals(DE, EE),
                         lessThan(FE, bigintLiteral(100)),
                         equals(AE, DE),
-                        equals(BE, EE)));
+                        equals(BE, EE),
+                        lessThan(BE, EE)));
```
Comment (kokosing) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/TestEffectivePredicateExtractor.java]:
```
          equals(BE, EE),
          lessThan(BE, EE)
```
This looks like `false`.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/EffectivePredicateExtractor.java [hunk 2]
```diff
@@ -231,11 +231,12 @@ public Expression visitJoin(JoinNode node, Void context)
...
-                            .add(pullExpressionThroughSymbols(leftPredicate, node.getOutputSymbols()))
-                            .add(pullExpressionThroughSymbols(rightPredicate, node.getOutputSymbols()))
-                            .addAll(pullExpressionsThroughSymbols(joinConjuncts, node.getOutputSymbols()))
-                            .build());
+                    return pullExpressionThroughSymbols(combineConjuncts(ImmutableList.<Expression>builder()
+                            .add(leftPredicate)
+                            .add(rightPredicate)
+                            .add(combineConjuncts(joinConjuncts))
+                            .add(node.getFilter().orElse(TRUE_LITERAL))
+                            .build()), node.getOutputSymbols());
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/EffectivePredicateExtractor.java]: I would extract variable out of:
```
combineConjuncts(ImmutableList.<Expression>builder()
                            .add(pullExpressionThroughSymbols(leftPredicate, node.getOutputSymbols()))	                            .add(leftPredicate)
                            .add(pullExpressionThroughSymbols(rightPredicate, node.getOutputSymbols()))	                            .add(rightPredicate)
                            .addAll(pullExpressionsThroughSymbols(joinConjuncts, node.getOutputSymbols()))	                            .add(combineConjuncts(joinConjuncts))
                            .build());	                            .add(node.getFilter().orElse(TRUE_LITERAL))
                            .build()
```

File: presto-main/src/main/java/com/facebook/presto/sql/planner/EffectivePredicateExtractor.java [hunk 3]
```diff
@@ -231,11 +231,12 @@ public Expression visitJoin(JoinNode node, Void context)
```
[the remark this answers is not in the document]
Comment (sopel39) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/EffectivePredicateExtractor.java]: You mean here? Very basic one:
```
SELECT c FROM t1 inner join t2 on (t1.a IN (1, 2, 3) AND t1.a = t2.b) inner join t3 on t2.b = t3.c
```
`t1` joined with `t2` won't have `a` as an output symbol. Therefore predicate `c IN (1,2,3)` won't propagate (without this change) during first predicate pushdown rotation.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/EffectivePredicateExtractor.java (hunk 2 above)
Comment (sopel39) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/EffectivePredicateExtractor.java]: Something is wrong with that code snippet.

File: presto-main/src/test/java/com/facebook/presto/sql/planner/TestEffectivePredicateExtractor.java (hunk 1 above)
Comment (sopel39) [reply presto-main/src/test/java/com/facebook/presto/sql/planner/TestEffectivePredicateExtractor.java]: That's true, but `EffectivePredicateExtractor` is not that clever right now. I will change to `lessOrEqual`...

File: presto-main/src/test/java/com/facebook/presto/sql/planner/TestEffectivePredicateExtractor.java [hunk 4]
```diff
@@ -458,7 +458,102 @@ public void testInnerJoin()
...
+                Optional.empty(),
+                Optional.empty(),
+                Optional.empty(),
+                Optional.empty());
+
+        Expression effectivePredicate = effectivePredicateExtractor.extract(node);
+
+        assertEquals(
+                normalizeConjuncts(effectivePredicate),
+                normalizeConjuncts(equals(DE, bigintLiteral(10))));
```
[the remark this answers is not in the document]
Comment (sopel39) [reply presto-main/src/test/java/com/facebook/presto/sql/planner/TestEffectivePredicateExtractor.java]: It was for `AE` (when `AE` is output symbol)

**Pull request merged**